### PR TITLE
Refactor input and output data of PaymentMethodDelegates

### PR DESCRIPTION
**Pull request by OscarSpruit**

### CODE REVIEW

File: bacs/src/main/java/com/adyen/checkout/bacs/BacsDirectDebitConfirmationView.kt [hunk 1]
```diff
@@ -51,7 +51,7 @@ internal class BacsDirectDebitConfirmationView @JvmOverloads constructor(
         this.localizedContext = localizedContext
         initLocalizedStrings(localizedContext)
 
-        bacsDelegate.outputData?.let {
+        bacsDelegate.outputData.let {
```
Comment (jreij) [inline bacs/src/main/java/com/adyen/checkout/bacs/BacsDirectDebitConfirmationView.kt:54]: Minor: maybe `with(bacsDelegate.outputData)` would be more suitable here?

File: googlepay/src/test/java/com/adyen/checkout/googlepay/DefaultGooglePayDelegateTest.kt [hunk 2]
```diff
@@ -30,41 +31,31 @@ import java.util.Locale
...
 
         @Test
         fun `payment data in null, then an exception should be thrown`() {
             assertThrows(CheckoutException::class.java) {
-                delegate.onInputDataChanged(GooglePayInputData(null))
-            }
-        }
-
-        @Test
-        fun `everything is good, then output data should be propagated`() = runTest {
```
Comment (jreij) [inline googlepay/src/test/java/com/adyen/checkout/googlepay/DefaultGooglePayDelegateTest.kt:57]: This test is totally removed? 🤔

File: online-banking-core/src/test/java/com/adyen/checkout/onlinebankingcore/DefaultOnlineBankingDelegateTest.kt [hunk 3]
```diff
@@ -90,47 +74,15 @@ internal class DefaultOnlineBankingDelegateTest(
...
+                with(expectMostRecentItem()) {
                     assertEquals("test", selectedIssuer?.name)
                     assertEquals("id", selectedIssuer?.id)
                     assertTrue(isValid)
                 }
             }
         }
-
-        @Test
-        fun `selectedIssuer is null, then component state should be invalid`() = runTest {
```
Comment (jreij) [inline online-banking-core/src/test/java/com/adyen/checkout/onlinebankingcore/DefaultOnlineBankingDelegateTest.kt:107]: Test completely removed

File: online-banking-core/src/test/java/com/adyen/checkout/onlinebankingcore/DefaultOnlineBankingDelegateTest.kt [hunk 4]
```diff
@@ -90,47 +74,15 @@ internal class DefaultOnlineBankingDelegateTest(
...
-
-                with(requireNotNull(expectMostRecentItem())) {
-                    assertEquals("", data.paymentMethod?.issuer)
-                    assertFalse(isValid)
-                }
-            }
-        }
-
-        @Test
-        fun `selectIssuer is valid, then component state should be valid`() = runTest {
```
Comment (jreij) [inline online-banking-core/src/test/java/com/adyen/checkout/onlinebankingcore/DefaultOnlineBankingDelegateTest.kt:121]: Test completely removed

File: blik/src/main/java/com/adyen/checkout/blik/BlikDelegate.kt [hunk 5]
```diff
@@ -15,23 +15,18 @@ import com.adyen.checkout.components.ui.ViewProvidingDelegate
...
+    PaymentMethodDelegate,
     ViewProvidingDelegate {
 
     val configuration: BlikConfiguration
 
-    val inputData: BlikInputData
-
     val outputData: BlikOutputData?
 
     val outputDataFlow: Flow<BlikOutputData?>
```
Comment (jreij) [inline blik/src/main/java/com/adyen/checkout/blik/BlikDelegate.kt:25]: `outputData` and `outputDataFlow` should not be nullable

**Comment by jreij** [comment]
Same comment for `outputData` and `outputDataFlow` but in Await, Voucher and QR code.

File: bacs/src/main/java/com/adyen/checkout/bacs/DefaultBacsDirectDebitDelegate.kt [hunk 6]
```diff
@@ -27,14 +27,18 @@ internal class DefaultBacsDirectDebitDelegate(
...
     private val _outputDataFlow = MutableStateFlow(createOutputData())
     override val outputDataFlow: Flow<BacsDirectDebitOutputData> = _outputDataFlow
 
+    override val outputData get() = _outputDataFlow.value
+
     private val _componentStateFlow = MutableStateFlow<BacsDirectDebitComponentState?>(null)
     override val componentStateFlow: Flow<BacsDirectDebitComponentState?> = _componentStateFlow
 
+    init {
+        createComponentState(outputData)
```
Comment (jreij) [inline bacs/src/main/java/com/adyen/checkout/bacs/DefaultBacsDirectDebitDelegate.kt:39]: Why do we need this on init?

File: blik/src/main/java/com/adyen/checkout/blik/DefaultBlikDelegate.kt [hunk 7]
```diff
@@ -28,17 +28,21 @@ internal class DefaultBlikDelegate(
 
     private val inputData: BlikInputData = BlikInputData()
 
-    private val _outputDataFlow = MutableStateFlow<BlikOutputData?>(null)
-    override val outputDataFlow: Flow<BlikOutputData?> = _outputDataFlow
+    private val _outputDataFlow = MutableStateFlow(createInitialOutputData())
```
Comment (jreij) [inline blik/src/main/java/com/adyen/checkout/blik/DefaultBlikDelegate.kt:31]: Can you use `createOutputData()` here?

File: voucher/src/main/java/com/adyen/checkout/voucher/VoucherView.kt [hunk 8]
```diff
@@ -90,6 +92,12 @@ internal class VoucherView @JvmOverloads constructor(
 
     override fun getView(): View = this
 
+    private fun loadLogo(paymentMethodType: String?) {
```
Comment (jreij) [inline voucher/src/main/java/com/adyen/checkout/voucher/VoucherView.kt:95]: Double checking now, I think I removed these 2 methods by mistake while refactoring:
```kotlin


    private fun launchDownloadIntent() {
        val url = component.outputData?.downloadUrl ?: return
        val intent = CustomTabsIntent.Builder()
            .setShowTitle(true)
            .setToolbarColor(ThemeUtil.getPrimaryThemeColor(context))
            .build()
        intent.launchUrl(context, Uri.parse(url))
    }

    private fun loadLogo(paymentMethodType: String?) {
        if (!paymentMethodType.isNullOrEmpty()) {
            imageLoader.load(paymentMethodType, binding.imageViewLogo, LogoApi.Size.MEDIUM)
        }
    }
```

Should we re-add the other one as well?

Comment (OscarSpruit) [reply voucher/src/main/java/com/adyen/checkout/voucher/VoucherView.kt:95]: Yes, because now there is no way to download the pdf anymore 😆 I will add it

**Review by jreij — APPROVED** [review]

**Pull request merged**